### ISSUE TITLE
Provide dedicated skin and hair shaders for creating realistic human characters

### ISSUE THREAD

**Issue by GeorgeS2019** [godotengine/godot-proposals#6900]
### Describe the project you are working on

An early concept of the feasibility of using Godot for creating realistic character

### Describe the problem or limitation you are having in your project

As we can expect more users would be interested in creating 3D games with a more realistic style, it'd be better if we can provide dedicated shaders to render human skin and hairs.

===>  This is **not my proposal** but simply [bringing **old (2019) suggestion** here](https://github.com/godotengine/godot/issues/27283)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

As I described in another issue [What it takes to create a photorealistic character](https://github.com/godotengine/godot/issues/26891), there are many obstacles that prevent one from creating a realistic human character in Godot and the lack of such dedicated shaders is one of them.

Moreover, a human skin material within an environment with a GI probe would either look completely dry or soaked in oil, depending on how it's shadowed, thus making it practically impossible to achieve even a remotely realistic result as is described in https://github.com/godotengine/godot/issues/19561.

And if we can accept what is discussed in that issue, it's apparently an expected behavior and we would need a dedicated skin shader to fix such a problem, so I think this would give this issue a bit more of importance than it would otherwise.

Currently, both Unity and Unreal provide their users such dedicated shaders to create a character with decent-looking skin and hairs by default. In comparison, we give our users a generic spatial material with which they can spend much time to make a plastic-looking skin that looks completely broken when it's shadowed, and they wouldn't find a clear remedy for the problem unless they are well versed in writing custom shaders (that is, provided https://github.com/godotengine/godot/issues/19561 can be actually solved by writing a shader, without any engine modification).

Eventually, we may be able to provide our users something like [this](https://docs.unrealengine.com/en-us/Resources/Showcases/PhotorealisticCharacter), so they can produce photorealistic results.

But in the meanwhile, I think even a much more primitive shader that can at least mitigate such a problem like https://github.com/godotengine/godot/issues/19561 would help those who try to create games with a more realistic style with Godot for now.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

![image](https://github.com/godotengine/godot-proposals/assets/49812372/fd52d804-7b0c-44a3-8db1-932ef47ce069)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No

### Is there a reason why this should be core and not an add-on in the asset library?

> This is add-on to create realistic human skin and hairs (revised based on feedback)

One of the [main feedback Juan received](https://godotengine.org/article/godot-at-gdc-2023/) from his presence at [Game Developers Conference](https://gdconf.com/) (GDC), on March 20-24 2023 was that there is a strong demand for improving Godot4 rendering.

I think this is a strong justification for why such photo-realistic rendering needs to go into ONE of its Future Core Features.

-----

_Further references:_

[Photo Realistic Graphics in Godot? Yes!](https://www.youtube.com/watch?v=tgK_pFdjFRA)

![image](https://github.com/godotengine/godot-proposals/assets/49812372/573ca698-e76f-4cdf-b2e2-40be55f67e29)

- [Godot Realistic Graphics](https://www.reddit.com/r/godot/comments/mwx6ju/godot_realistic_graphics/)
- [Photorealism in Godot 4 !](https://youtu.be/d_BJ9RN6PLE)
-  [Photorealistic demo made in a few hours in Godot 4](https://www.reddit.com/r/godot/comments/130e009/photorealistic_demo_made_in_a_few_hours_in_godot_4/)

- [Godot Photorealistic Graphics](https://www.youtube.com/watch?v=a38LzdOi6KI)
![image](https://github.com/godotengine/godot-proposals/assets/49812372/b8f6ba43-9716-4189-8038-7050d6a2d773)

- [Realistic human skin practice](https://www.therookies.co/projects/3289)
![image](https://github.com/godotengine/godot-proposals/assets/49812372/c89c18ee-e01f-4c2f-8b20-cb53ade058cf)

**Comment by GeorgeS2019** [comment]
[Photo Realistic Graphics in Godot? Yes!](https://www.youtube.com/watch?v=tgK_pFdjFRA)

![image](https://github.com/godotengine/godot-proposals/assets/49812372/573ca698-e76f-4cdf-b2e2-40be55f67e29)

**Comment by GeorgeS2019** [comment]
[Godot Realistic Graphics](https://www.reddit.com/r/godot/comments/mwx6ju/godot_realistic_graphics/)

**Comment by Zireael07** [comment]
> Is there a reason why this should be core and not an add-on in the asset library?
This is add-on to create realistic human skin and hairs

If you proposed an addon, then this is not the place for this.

If this is supposed to go into core, then this is not sufficient explanation why this should be core

**Comment by AThousandShips** [comment]
Adding to the argument that this is addon territory rather than core, beyond the aim that Godot should be lean, is that it places the burden of responsibility to not be caught in biases about human diversity on the development of Godot, which is a major and important thing that requires a lot of work

**Comment by GeorgeS2019** [comment]
> The human diversity 

could be addressed by the add-on and the core human realistic rendering could be a future Core feature.

**Comment by AThousandShips** [comment]
What do you mean by "core human realistic rendering"?

**Comment by smix8** [comment]
This can basically be solved with a documentation page and some shader code examples like Godot documentation already has for some other shader stuff that can not be solved with just basic StandardMaterial3D.

A lot that goes into a proper "realistic" look is texture and mesh model work and good light setup, the shader is only one part of that.

**Comment by GeorgeS2019** [comment]
@AThousandShips  

To be honest, I have not looked into the details of how Unreal and Unity are doing that yet. Likewise for other 3D engines like e.g. reallusion.com

However, I totally agree with the human diversity aspect.

**Comment by GeorgeS2019** [comment]
@Zireael07 For me, the priority there is a way to coordinate Godot community interest to contribute to this feature, I let the community decide

The most important aspect, it helps the iteration of improving this feature as Godot is still lacking demos and examples in this aspect.

**Comment by AThousandShips** [comment]
I believe that Godot should provide the base but not the specifics for high level functionality like specific photorealism, Godot should be flexible enough to allow users to do this, while not adding the complexity of providing the specific implementation, in other words work towards making Godot better for photorealism on one hand, and adding tools or bundled resources for doing specific things on the other

The former liberates and creates room for what you want to accomplish, but the later risks restricting and forcing you to make your own anyway if it doesn't fit

**Comment by AThousandShips** [comment]
(Also consider the fact that all the examples of photorealistic humans you have provided yourself are of light skin)

**Comment by GeorgeS2019** [comment]
@AThousandShips 
>  light skin

I respect inclusive

This is MORE Important reasons that Godot needs to demonstrate high quality rendering on **NON-light skin** characters

**Comment by AThousandShips** [comment]
It *has* to be inclusive, anything else is unacceptable, and my concern here is the harm that can be caused if Godot makes specific decisions in creating resources for realistic rendering of humans based on these biases, but agreed the ability of the rendering system to accomplish this is also important, but as this is about dedicated shaders for this the risk of that bias should be the focus imo

**Comment by GeorgeS2019** [comment]
Once more, Godot is the 8th most popular open source GitHub project. It is MEANT for ALL

This shall address different ethnicities like MakeHuman

**Comment by smix8** [comment]
I though this topic was about hair and skin shader, which has nothing at all to do with your inclusivity concerns.

The things you are concerned about is what kind of body shape the character meshes have or what albedo colors the used textures have, which again, has nothing at all to do with how to make a proper hair or skin shader.

**Comment by GeorgeS2019** [comment]
@smix8 
Agree. However, we need to respect the diversity of Godot users and their concerns so nothing is left out during planning and design

**Comment by AThousandShips** [comment]
Maybe I misunderstood this but this proposal sounded like a way to create realistic human skin shaders out of the box, and MetaHuman was given as an example which seems to be providing that feature, i.e. creating from scratch this feature, that's where the concerns are

**Comment by AThousandShips** [comment]
Also I'm not convinced that these aspects of diversity doesn't apply even just with the basic aspects of the parameters of plain shaders without making any input from the engine on albedo, I am not informed enough but I don't think it can be dismissed either

**Comment by YuriSizov** [comment]
Thanks for showing the interest in this proposal, everyone, but I would like to remind you that this is a discussion platform, not a chat. Please avoid posting several messages in quick succession, and use the edit button instead. Every post you make creates a notification for everyone subscribed to the repository (or the issue), and several disjointed comments are harder to read than a single concise one.

I would also urge you to take a bit of time before replying, so your answer doesn't require one or two more addendums, and so the other party doesn't feel hurried by your immediate response to argue back. If you want to have a more live discussion on the matter, I invite you to the Godot contributors chat at https://chat.godotengine.org.

**Comment by mrjustaguy** [comment]
My 2 cents on this

1) We don't need a dedicated Skin Shader, Base Material works just fine and can fairly easily get the right values for proper skin, and this can be documented for those that don't get proper Texture Maps for skin bundled with the humans, such as is the case with MakeHuman mentioned above
2) The Main issue that's causing the skin to look wrong is a problem inherent with raster, and that's reflections, they look wrong in most instances. This cannot be simply fixed with a dedicated shader, only mitigated by artists dealing with environment lighting. A true solution would be to do Ray Tracing, as that's the only way to get proper reflections
3) Only a Hair Shader makes sense, something along the lines of TressFX https://gpuopen.com/tressfx/ as hair does need to be simulated and card hair is typically horrible and quite obvious
4) I'd understand a Dedicated Human Node in the editor, for things like Character creators and for having a Node that handles realistic motion, however both don't have license compatible bases to work off and are massive undertakings each in their own right, and is something likely to start out as an addon if ever

**Comment by Zireael07** [comment]
@mrjustaguy I never could get skin materials working passably well so it would need more documentation indeed

**Comment by yosoyfreeman** [comment]
I think it can be done with standard PBR. 

You can use visual shading to create the layers required for your needs in relatively intuitive way. 

Here is a super good example: An incredible looking skin shader made with common pbr workflow inside unity optimized even for mobile devices. 

You can take a look at the properties and see some of the layers and configuration needed for a setup like that.

But i still think official demos about this topic would help everyone.

**Comment by GeorgeS2019** [comment]
[WIP] Godot Demo Project for realistic skin

[Godot PBR tutorial - Blender 2.93](https://www.youtube.com/watch?v=gcft7kF3lYk)

[Can't import shaders/textures to Godot from Blender 2.8](https://gamedev.stackexchange.com/questions/185081/cant-import-shaders-textures-to-godot-from-blender-2-8)

**Comment by and-rad** [comment]
I'm gonna shamelessly plug one of the projects I'm working on where I explore Godot's capabilities for rendering digital humans:

https://github.com/RoboPoets/digital_human

So far I've only gotten to the eyes, but the current implementation is close to on par with Unreal's and Unity's eye shaders. I agree that this is nothing that needs to be added to core and instead should be done via plugins. Hair might turn out to be an exception if the goal is to create a strand-based system. This is definitely something I want to look into and it could require core modifications, but I'm not in a position to tell yet.

Until then I think the best this proposal can hope for is a discussion about whether an additional demo project for this topic should be created and if so, what it should contain.

**Comment by GeorgeS2019** [comment]
@and-rad  
What is your view on human skin? Do you have a project where you take e.g. Unity3D or reallusion texture and shading attempting to do that in Godot4?

The ideal demo project will be like MetaHuman

There will be animations, both skeletal and facial blendshape that mimic the way MetaHuman move and express his or her **Inner thinking** when alone.

When we view this character in Godot, we wonder what he or she is thinking!

**Comment by and-rad** [comment]
Don't overhype MetaHuman. The system is neither the best nor the most efficient at what it does. The polycount is way higher than it needs to be and the number of joints is just completely insane. Whenever people go on about how good Metahumans are, I like to point at [this tweet](https://twitter.com/wasim_cg/status/1279288006734561281?lang=en) from one of the animators working on TLOU2.

> The ideal demo project will be like MetaHuman

What exactly do you mean here? MetaHuman is not an Unreal Engine project, it is a software that generates humanoid characters that you can import into Unreal Engine. MetaHuman doesn't really have anything to do with the quality of the shaders used. You can slap a realistic skin shader on a cube and it's going to look just as good.

> The Facial Blendshapes will be based on [ArKit 52 blendshape Industry standards ](https://github.com/google/mediapipe/issues/3421#issuecomment-1483426344)and nothing ELSE.

This is waaay too dogmatic. Not everyone needs this kind of fidelity and some are going to need even more. This insistence of yours on the One True Way™ to do these things is just not going to work within the greater Godot ecosphere.

Also, I'd like to point out that this proposal (written by you) is only about shaders and rendering, not about animation features.

**Comment by GeorgeS2019** [comment]
@and-rad 

Apology. 

I am jumping too many steps ahead

Yes, the The polycount is way higher, so as that of reallusion. 
Yes, that is not the priority. We need to just demo, the skin as realistic as possible and if possible there is a way of dealing with hair as well
Yes, the priority will be FIRST a realistic skin, that itself is missing and we need we who know to guide us. 

Yes, we reserve animation outside this discussion. The animation I suggested is less a challenge than the realistic skin.

**Comment by and-rad** [comment]
That is certainly a more reasonable scope. There's two things I like to clarify though.

> Yes, the The polycount is way higher, so as that of reallusion.

Maybe I'm reading this wrong, but are you saying that both MetaHumans and Reallusion characters have a similar polycount? Because that's not correct. CC4 characters have about 20K polys total, while a MetaHuman at LOD0 has 30k polys in the face (not the head, the *face*) alone. Which just goes to show again that you really don't need these enormous polycounts to get good realism.

> Yes, we defer animation outside this discussion. The animation I suggested is less a challenge than the realistic skin.

Not even close. Realistic skin rendering is going to be way easier to achieve. It's basically already a solved problem with Godot's current rendering capabilities, including stuff like wrinkle maps and blood flow. Someone just has to write a good, tweakable shader. Creating animations on the other hand, even just as few as 50 blend shapes for the face, is going to be so much more work.

The way I see it, this is a rough hierarchy from easiest to hardest:

1. Skin
2. Eyes
3. Hair
4. Animation

Eyes include some complex math, but there are enough papers out there to get by. Hair is technologically challenging if you want to go beyond cards, and resources are sparse once you go that way. Animation is artistically challenging because the smallest mistake can easily land you deep in the uncanny valley. It also takes so much time and you need skilled modelers to create the shapes, and a well-thought-out workflow to be able to reuse them across different character models.

**Comment by GeorgeS2019** [comment]
@and-rad  
There is no problem if anyone say I do not know anything. 
Most important for me that this demo project moves ahead. That is key for the community :-)

**Comment by yosoyfreeman** [comment]
Yeap. Yesterday i spend a considerable ammount of time trying to provide an answer with some value, but OP posted and deleted or edited things so many time that i desisted after an hour. Including posting messages "Only for x user that i will delete soon cause godot admins does not allow feedback" and i surrendered after that sentence. Please close this.

**Comment by GeorgeS2019** [comment]
@yosoyfreeman 

![image](https://github.com/godotengine/godot-proposals/assets/49812372/b48b7a49-c1e3-4f9f-a3a2-cc0e138e73fd)

**Comment by AThousandShips** [comment]
Doesn't explain removing comments from a dialog, the comment was about not spamming with *new* comments, the reason your comments were deleted before was that they were moved to the proposal body to make it more navigable

All you did was remove all the context of our comments and break any kind of logic to an ongoing discussion

**Comment by yosoyfreeman** [comment]
I'm sorry but I'm unable to keep with this conversation.  Now there is no context, all our answers won't help anyone in the future reading this. You just marked this as closed and eleven minutes later continued with the conversation.

I just will say that @and-rad work is a demonstration of current Godot capabilities and he/she even said does not find any that needed to be added as core.

That magnificent work is awesome, yes, but releasing it or not is not our choice but the choice of the artist who made them. It may be a plugin. It may be a paid asset. It may be a demo. Or it may be nothing cause he or she wants to keep their work.  I feel this is more asking about game assets that is a Godot feature request. 

I'm not trying to be rude, I'm just unable to understand what we are even discussing at this point.

**Comment by Zireael07** [comment]
@yosoyfreeman there's nothing left to discuss here since OP deleted their comments thus rendering this discussion/conversation a mess

**Comment by matt-seaton** [comment]
As someone who is working on a character creator for Godot (see here: https://github.com/NitroxNova/Humanizer) I think there are bigger problems than photorealistic shaders.

**Comment by GeorgeS2019** [comment]
We had Godot 4 makehuman before, deleted due to commercial reason
Photorealistic skin is the key problem

**Comment by matt-seaton** [comment]
Our plugin is makehuman for godot and it's not deleted.

You can't have photorealism if you can't write a working skin shader due to the skinning pipeline.

**Comment by GeorgeS2019** [comment]
The deleted is Godot4  c# version of gdscript Godot 3 old version of https://github.com/vikram3/Godot4CharacterCreator

**Comment by GeorgeS2019** [comment]
The skin shader provided by makehuman is no way close to CC3

**Comment by matt-seaton** [comment]
I know about Go-Makehuman-dot.  Ours is a much more advanced than that and I'm telling you there are fundamental roadblocks in Godot shaders and the skinning pipeline that have to be addressed before we can make progress here.  

I don't care about makehuman's skin shader.  It's completely irrelevant.  It's not compatible with godot anyway.

**Comment by GeorgeS2019** [comment]
>fundamental roadblocks in Godot shaders and the skinning pipeline that have to be addressed

I am not the expert
If U know how to articulate the missing parts, please list them and we iterate through to get other more experts to involve
 Thx for pointing out

**Comment by GeorgeS2019** [comment]
For trimesh, I heard someone spoke before. Join V-Sekai discord

Speak to iyamma

**Comment by matt-seaton** [comment]
I already did in my original post.  The main problem is the lack of accessibility to data from the gpu.  We need to either be able to implement a fragment shader and grab the resulting texture maps (and preferably skinned vertex coordinates) from the gpu, or we need to have access to un-skinned vertex positions in the vertex shader, preferably the former.

**Comment by GeorgeS2019** [comment]
https://youtu.be/ft--lVFpx4E?si=yGxtPq9NPVmSmBNM

**Comment by matt-seaton** [comment]
There is no body.  There are probably no seams.  If you only want headless people maybe that can work for you.  We also have face tracking animations.  Not a big deal.

**Comment by Calinou** [comment]
> 1. For staticbodies (e.g. humanoid statues) we cannot bake a posed mesh to generate a trimesh collider. That data exists only on the gpu and is inaccessible on the cpu side.

This is available in 4.3 thanks to https://github.com/godotengine/godot/pull/76725. You can bake a mesh from a blendshape mix, then generate a trimesh collision shape based on the baked mesh.

**Comment by matt-seaton** [comment]
> > 1. For staticbodies (e.g. humanoid statues) we cannot bake a posed mesh to generate a trimesh collider. That data exists only on the gpu and is inaccessible on the cpu side.
> 
> For animations, there's an open PR: [godotengine/godot#85018](https://github.com/godotengine/godot/pull/85018)
> 
> For blend shapes, this is available in 4.3 thanks to [godotengine/godot#76725](https://github.com/godotengine/godot/pull/76725). You can bake a mesh from a blendshape mix, then generate a trimesh collision shape based on the baked mesh.

Good to know.  For us, we are using makehuman target files, so we already have the mesh data for blendshapes.  We really need the posed skeleton data for that.
